### DISCUSSION TITLE
Help: Remap Ctrl-n to "switch to n-th buffer"

### DISCUSSION THREAD

**Discussion by tf42src** [LazyVim/LazyVim#3746] [Q&A] [answered]
I have this code in my **keymaps.lua**

```
local defaults = { noremap = true, silent = true }

-- function to switch to the nth buffer
_G.goto_buffer = function(n)
  return function()
    vim.api.nvim_command('buffer ' .. n)
  end
end

-- map Ctrl-1 to Ctrl-9 to switch to buffer 1 to 9
for i = 1, 9 do
  vim.api.nvim_set_keymap('n', '<C-' .. i .. '>', ':lua goto_buffer(' .. i .. ')()<CR>', defaults)
end
```

The function is called when Ctrl-n is pressed, but the buffer numbers are not sequential. Executing `:buffers` shows me my four files have the buffer ids: 3, 22, 75, 77.

How can I remap my keys so that I can easily jump to a specific buffer using Ctrl-n ?

**Comment by dpetka2001** [comment]
Something like the following could work
```lua
local buffers = {}
for _, buf in ipairs(vim.api.nvim_list_bufs()) do
  if vim.api.nvim_buf_is_valid(buf) and vim.bo[buf].buftype == "" then
    table.insert(buffers, buf)
  end
end

for k, v in ipairs(buffers) do
  vim.keymap.set("n", "<M-" .. k .. ">", "<cmd>buffer " .. v .. "<cr>", { desc = "go to buffer" })
end
```
But I personally prefer setting a keymap with Bufferline (if you use it) to do the same
```lua
vim.keymap.set("n", "<leader>bb", function()
  require("bufferline").pick()
end, { desc = "Pick buffer" })
```

**Reply by tf42src** [reply]
I have added the bufferline solution. Will give it a try, it seems quite neat.

Nevertheless, it would be great to get Ctrl-n working as well (I assume I will just be faster - and I don't want to give up). 

The solution you posted doesn't seem to work out of the box. I assume the for loop needs to run inside a function? (because the number of buffers changes all the time). I got it working with a mix of your code and mine:

```
_G.goto_buffer = function(n)
  local counter = 1
  for _, buf in ipairs(vim.api.nvim_list_bufs()) do
    if vim.api.nvim_buf_is_valid(buf) and vim.bo[buf].buftype == "" then
      if counter == n then
        vim.api.nvim_command("buffer " .. buf)
        break
      end
      counter = counter + 1
    end
  end
end

-- map Ctrl-1 to Ctrl-9 to switch to buffer 1 to 9
for i = 1, 9 do
  vim.api.nvim_set_keymap("n", "<C-" .. i .. ">", ":lua goto_buffer(" .. i .. ")()<CR>", defaults)
end
```

It works. The only annoying part left is the following error message (which does not occur if I type `:lua goto_buffer(1)`). Do I need to return something from the function?

![Screenshot from 2024-06-20 23-50-11](https://github.com/LazyVim/LazyVim/assets/9712861/f10b14f9-ebdf-4c6c-9d64-14d501024320)

**Reply by tf42src** [reply]
I am struggling to understand what `nvim_list_bufs()` does based on the [documentation](https://neovim.io/doc/user/api.html#nvim_list_bufs()). What is a "buffer handle"? I am asking because replacing the wildcard in the loop with `counter` didn't have the same result as adding and incrementing the counter manually.

**Edit:**
Seems there are buffers which get filtered out in the if-clause and hence I cannot use the ipairs index.

**Reply by dpetka2001** [reply]
A buffer handle is the the buffer id of each buffer. When you do `:buffers` you get some numbers and partial buffer names. Those numbers are the buffer ids/handlers.

**Reply by dpetka2001** [reply, marked as answer]
I put the following in `~/.config/nvim/lua/config/autocmds.lua` and it seems to do what you want
```lua
vim.api.nvim_create_autocmd({ "BufReadPost", "BufNewFile" }, {
  group = vim.api.nvim_create_augroup("test_buffers", {}),
  desc = "test buffers",
  callback = function()
    local buffers = {}
    for _, buf in ipairs(vim.api.nvim_list_bufs()) do
      if vim.api.nvim_buf_is_valid(buf) and vim.bo[buf].buftype == "" then
        table.insert(buffers, buf)
      end
    end

    for k, v in ipairs(buffers) do
      vim.keymap.set("n", "<M-" .. k .. ">", "<cmd>buffer " .. v .. "<cr>", { desc = "go to buffer" })
    end
  end,
})
```
Test it out.

**Reply by dpetka2001** [reply]
Still I believe this is not needed with `bufferline.pick`. I just tested it for educational purposes.

**Reply by tf42src** [reply]
> I put the following in `~/.config/nvim/lua/config/autocmds.lua` and it seems to do what you want
> 
> ```lua
> vim.api.nvim_create_autocmd({ "BufReadPost", "BufNewFile" }, {
>   group = vim.api.nvim_create_augroup("test_buffers", {}),
>   desc = "test buffers",
>   callback = function()
>     local buffers = {}
>     for _, buf in ipairs(vim.api.nvim_list_bufs()) do
>       if vim.api.nvim_buf_is_valid(buf) and vim.bo[buf].buftype == "" then
>         table.insert(buffers, buf)
>       end
>     end
> 
>     for k, v in ipairs(buffers) do
>       vim.keymap.set("n", "<M-" .. k .. ">", "<cmd>buffer " .. v .. "<cr>", { desc = "go to buffer" })
>     end
>   end,
> })
> ```
> 
> Test it out.

Works like a charm! And I also learned about auto commands, thank you!

_For potential others, need to replace `M-` with `C-` to use ctrl instead of modifier (alt)._

**Reply by tf42src** [reply]
@dpetka2001 what is the autocmd hook for "move buffer" (e.g. `<leader>[B`)? ;)

**Edit:**
Actually I don't think this would work either as the IDs remain the same. So the question would be how do I get the current buffer position in the list of buffers?

**Reply by tf42src** [reply]
> I have added the bufferline solution. Will give it a try, it seems quite neat.
> 
> Nevertheless, it would be great to get Ctrl-n working as well (I assume I will just be faster - and I don't want to give up).
> 
> The solution you posted doesn't seem to work out of the box. I assume the for loop needs to run inside a function? (because the number of buffers changes all the time). I got it working with a mix of your code and mine:
> 
> ```
> _G.goto_buffer = function(n)
>   local counter = 1
>   for _, buf in ipairs(vim.api.nvim_list_bufs()) do
>     if vim.api.nvim_buf_is_valid(buf) and vim.bo[buf].buftype == "" then
>       if counter == n then
>         vim.api.nvim_command("buffer " .. buf)
>         break
>       end
>       counter = counter + 1
>     end
>   end
> end
> 
> -- map Ctrl-1 to Ctrl-9 to switch to buffer 1 to 9
> for i = 1, 9 do
>   vim.api.nvim_set_keymap("n", "<C-" .. i .. ">", ":lua goto_buffer(" .. i .. ")()<CR>", defaults)
> end
> ```
> 
> It works. The only annoying part left is the following error message (which does not occur if I type `:lua goto_buffer(1)`). Do I need to return something from the function?
> 
> ![Screenshot from 2024-06-20 23-50-11](https://private-user-images.githubusercontent.com/9712861/341437790-f10b14f9-ebdf-4c6c-9d64-14d501024320.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg5Mzk2NTUsIm5iZiI6MTcxODkzOTM1NSwicGF0aCI6Ii85NzEyODYxLzM0MTQzNzc5MC1mMTBiMTRmOS1lYmRmLTRjNmMtOWQ2NC0xNGQ1MDEwMjQzMjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjFUMDMwOTE1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzgxYzAzNDFkZWJjODdkOGEwZjc3YTA1NTBhNTUwZmY5MWQ3YmE4YjFkMzA4MWIzYzRkMmE3N2M2N2RkOGVjYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.5PV909LebCKjTVO2K2rovqq9St-Yy_s-LZp-ySN6ABk)

To solve the error in the screenshot, I had to remove the extra `()` after the `goto_buffer` call, before the `<CR>`, like so:
`  vim.api.nvim_set_keymap("n", "<C-" .. i .. ">", ":lua goto_buffer(" .. i .. ")<CR>", defaults)`
